File: backend/services/card_numbers.py
import re
from typing import Optional

_DIGITS_RE = re.compile(r"^\d+$")


def normalize_card_number(value: object) -> str:
    """Normalize numeric card numbers so 44 and 044 compare equally."""
    text = "" if value is None else str(value).strip()
    if not text:
        return ""
    if _DIGITS_RE.fullmatch(text):
        return text.lstrip("0") or "0"
    return text.lower()


def card_number_matches(stored_number: Optional[str], requested_number: object) -> bool:
    stored = "" if stored_number is None else str(stored_number).strip()
    requested = "" if requested_number is None else str(requested_number).strip()
    if not stored or not requested:
        return False
    if stored == requested:
        return True
    return normalize_card_number(stored) == normalize_card_number(requested)

File: frontend/src/utils/cardNumbers.js
const DIGITS_RE = /^\d+$/

export function normalizeCardNumber(value) {
  const text = String(value ?? '').trim()
  if (!text) return ''
  if (DIGITS_RE.test(text)) return text.replace(/^0+/, '') || '0'
  return text.toLowerCase()
}

export function cardNumberMatches(cardNumber, query) {
  const stored = String(cardNumber ?? '').trim()
  const requested = String(query ?? '').trim()
  if (!stored || !requested) return false
  if (stored.toLowerCase() === requested.toLowerCase()) return true
  return normalizeCardNumber(stored) === normalizeCardNumber(requested)
}
